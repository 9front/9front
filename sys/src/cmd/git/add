#!/bin/rc -e
rfork ne
. /sys/lib/git/common.rc

gitup

flagfmt='r:remove'; args='file ...'
eval `''{aux/getflags $*} || exec aux/usage

add='tracked'
del='removed'
if(~ $remove 1){
	add='removed'
	del='tracked'
}
if(~ $#* 0)
	exec aux/usage

if(~ $add tracked)
	files=`$nl{walk -f $gitrel/$*}
if not
	files=`$nl{cd .git/index9/tracked/ && walk -f $gitrel/$*}

for(f in $files){
	if(! ~ `{cleanname $f} .git/*){
		addpath=.git/index9/$add/$f
		delpath=.git/index9/$del/$f
		mkdir -p `{basename -d $addpath}
		mkdir -p `{basename -d $delpath}
		# We don't want a matching qid, so that
		# git/walk doesn't think this came from
		# a checkout.
		if(! test -e $addpath)
			if(~ $add 'tracked' || test -e /mnt/git/HEAD/tree/$f)
				touch $addpath
		rm -f $delpath
	}
}
exit ''
