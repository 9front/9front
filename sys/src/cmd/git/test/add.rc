#!/bin/rc

. util.rc

rm -fr scratch
mkdir -p scratch/repo

echo @@ abs paths @@
@{
	rfork ne
	cd scratch/repo
	pwd=`{pwd}
	$G/init
	mkdir dir
	mkdir another
	mkdir more
	mkdir extra
	touch a b c
	touch dir/^(a b c)
	touch another/^(a b c)
	touch more/^(a b c)
	touch extra/^(a b c)
	$G/add a
	$G/add $pwd/b
	$G/add dir/a
	$G/add $pwd/dir/b
	$G/add more
	@{cd more && $G/add ../extra/a}
	@{cd more && $G/add $pwd/extra/b}
	$G/$O.fs
	$G/$O.walk > ../added
}

diff -c scratch/added add.expected >/dev/null || die wrong files
