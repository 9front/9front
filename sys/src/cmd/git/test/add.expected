A a
A b
A dir/a
A dir/b
A extra/a
A extra/b
A more/a
A more/b
A more/c
